feat(sagemaker): parallelize tag listing for better performance

sonofagl1tch opened the pull request:
### Context

The current implementation of the SageMaker service in Prowler fetches resource tags sequentially. For accounts with a large number of Models, Notebook Instances, Training Jobs, or Endpoint Configs, this results in blocking API calls that significantly slow down the execution of the provider scan.

### Description

This PR improves performance by parallelizing the tag retrieval process for SageMaker resources.

- Refactored [_list_tags_for_resource] in [sagemaker_service.py] to handle individual resources.
- Updated [__init__] to use `__threading_call__` for concurrent execution of tag retrieval.
- Added unit tests in [tests/providers/aws/services/sagemaker/sagemaker_service_tags_test.py].

### Steps to review

1.  **Code Review**: Check [prowler/providers/aws/services/sagemaker/sagemaker_service.py] to verify that [_list_tags_for_resource] is now called via `__threading_call__` for each resource list (`sagemaker_models`, `sagemaker_notebook_instances`, etc.).
2.  **Verification**:
    - Run the new unit tests:
      ```bash
      poetry run pytest tests/providers/aws/services/sagemaker/sagemaker_service_tags_test.py
      ```
    - Run existing tests to ensure no regressions:
      ```bash
      poetry run pytest tests/providers/aws/services/sagemaker/sagemaker_service_test.py
      ```

### Checklist

- Are there new checks included in this PR? **No**
    - If so, do we need to update permissions for the provider? N/A
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [ ] Review if backport is needed.
- [ ] Review if is needed to change the [Readme.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [x] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### UI
- [ ] All issue/task requirements work as expected on the UI
- [ ] Screenshots/Video of the functionality flow (if applicable) - Mobile (X < 640px)
- [ ] Screenshots/Video of the functionality flow (if applicable) - Table (640px > X < 1024px)
- [ ] Screenshots/Video of the functionality flow (if applicable) - Desktop (X > 1024px)
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/ui/CHANGELOG.md), if applicable.

#### API
- [ ] Verify if API specs need to be regenerated.
- [ ] Check if version updates are required (e.g., specs, Poetry, etc.).
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.


closes https://github.com/prowler-cloud/prowler/issues/9608

sonofagl1tch committed:
feat(sagemaker): parallelize tag listing for better performance

- Refactored _list_tags_for_resource to handle individual resources.
- Updated __init__ to use __threading_call__ for concurrent execution of tag retrieval.
- Added unit tests.

jfagoagas reviewed (changes requested):
That's a great improvement @sonofagl1tch 👏 Thanks for taking care of this. I left one comment regarding tests. 

Also, please: 
- Take a look at the CI because some linters are failing. 
- Add an entry into `prowler/CHANGELOG.md` for the next version.

Let us know if you need help with that. 🙌

sonofagl1tch committed:
test(sagemaker): consolidate tag tests and fix linting

- Merge sagemaker_service_tags_test.py into sagemaker_service_test.py as requested by reviewer.
- Fix flake8 linting errors (unused imports and shadowing).
- Apply black formatting to test files.
- Add entry to prowler/CHANGELOG.md for parallel tag retrieval.

sonofagl1tch committed:
fixing error: Run poetry run flake8 . --ignore=E266,W503,E203,E501,W605,E128 --exclude contrib,ui,api
  poetry run flake8 . --ignore=E266,W503,E203,E501,W605,E128 --exclude contrib,ui,api
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.12.12/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.12.12/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.12/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.12/x64/lib
<unknown>:272: SyntaxWarning: invalid escape sequence '\ '
<unknown>:187: SyntaxWarning: invalid escape sequence '\ '
<unknown>:1324: SyntaxWarning: invalid escape sequence '\.'
<unknown>:228: SyntaxWarning: invalid escape sequence '\d'
./prowler/providers/aws/services/sagemaker/sagemaker_service.py:25:1: W293 blank line contains whitespace
./prowler/providers/aws/services/sagemaker/sagemaker_service.py:37:1: W293 blank line contains whitespace
./prowler/providers/aws/services/sagemaker/sagemaker_service.py:207:5: E303 too many blank lines (2)
Error: Process completed with exit code 1.

jfagoagas committed:
fix: changelog order

jfagoagas committed:
Merge branch 'master' into prowler-sagemakerParallelCalls

jfagoagas reviewed (approved):
That's a great improvement @sonofagl1tch 👏

jfagoagas commented:
> @jfagoagas i moved the code and added the chagelog entry. I am not getting any linter errors on my side and i rebuilt everything in a new docker container and ran a scan to confirm it worked end to end. Can you let me know if there are still issues? if so, i am going to build fresh environment because I currently can't replicate the linter issues.

Everything is ready!
